### PR TITLE
use command line parser to handle user input

### PR DESCRIPTION
**Pull request by tschuerl**
Now using a popular command line parser to handle typeCart arguments. When a flag is raised, all arguments after the flag are added to list (ie. for command `typecart -r hello hi -d last` the parser will make the optional list `r= ["hello", "hi"]` and `d = ["last"]`). My logic is to see which flags contain arguments and then run appropriate functions (an empty list implies the flag wasn't called)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: TypeInjections/Tool/Tool.fs [hunk 1]
```diff
@@ -21,31 +21,38 @@ module Tool =
...
-        elif flag.Contains("-l") then
-            Console.WriteLine("***** git commit")
+    type options =
+        { [<Option('d', "delete", Required = false, HelpText = "use option to remove New and Old folders")>]
+          delete: bool
+          [<Option('l', "local", Required = false, HelpText = "[inputPath] [inputPath] [outputPath]")>]
+          local: seq<string>
+          [<Option('g', "git", Required = false, HelpText = "[old commit] [new commit] [outputPath]")>]
+          git: seq<string>
+          [<Option('r', "remote", Required = false, HelpText = "[github repo url] [new commit] [old commit] [OutputPath]")>]
```
Comment (mschlaipfer) [inline TypeInjections/Tool/Tool.fs]: We mix multiple words, lower camel case and upper camel case here, please be consistent

File: TypeInjections/Tool/Tool.fs [hunk 2]
```diff
@@ -21,31 +21,38 @@ module Tool =
...
-            Console.WriteLine("***** local project comparison")
-
-            if argvList.Length <> 4 then
-                failwith "usage: program -p OLD NEW OUTPUTFOLDER"
 
-        // -l [git commit hash] [git commit hash] [OutputPath]
-        elif flag.Contains("-l") then
-            Console.WriteLine("***** git commit")
+    type options =
+        { [<Option('d', "delete", Required = false, HelpText = "use option to remove New and Old folders")>]
```
Comment (mschlaipfer) [inline TypeInjections/Tool/Tool.fs]: Are `New` and `Old` the folder names? Then make it `the New and Old folders`. If not please lower case them.

Does this flag really just delete them? I don't think we should offer such functionality. Or does it delete existing ones before writing new ones? This is often called `force` (maybe this would be a more fitting name).
If it does more than just delete them, please reflect that in the help text.

File: TypeInjections/Tool/Tool.fs [hunk 3]
```diff
@@ -21,31 +21,38 @@ module Tool =
...
+          delete: bool
+          [<Option('l', "local", Required = false, HelpText = "[inputPath] [inputPath] [outputPath]")>]
```
Comment (mschlaipfer) [inline TypeInjections/Tool/Tool.fs]: What are the input paths for? Please be more descriptive in the help text.

File: TypeInjections/Tool/Tool.fs [hunk 4]
```diff
@@ -21,31 +21,38 @@ module Tool =
...
+          local: seq<string>
+          [<Option('g', "git", Required = false, HelpText = "[old commit] [new commit] [outputPath]")>]
```
Comment (mschlaipfer) [inline TypeInjections/Tool/Tool.fs]: Why does the `git` command take an `outputPath`?

File: TypeInjections/Tool/Tool.fs (hunk 1 above)
Comment (mschlaipfer) [reply TypeInjections/Tool/Tool.fs]: What's the difference between `remote` and `git`?

File: TypeInjections/Tool/Tool.fs [hunk 5]
```diff
@@ -21,31 +21,38 @@ module Tool =
...
+          git: seq<string>
+          [<Option('r', "remote", Required = false, HelpText = "[github repo url] [new commit] [old commit] [OutputPath]")>]
+          remote: seq<string>
+          [<Option(HelpText = "Prints all messages to standard output.")>]
+          verbose: bool }
 
-            if argvList.Length <> 4 then
-                failwith "usage: program -l OLD_COMMIT NEW_COMMIT OUTPUTFOLDER"
+    // based off command line flags, create and provide path for input/output folders
```
Comment (mschlaipfer) [inline TypeInjections/Tool/Tool.fs]: This comment doesn't help me

File: TypeInjections/Tool/Tool.fs [hunk 6]
```diff
@@ -21,31 +21,38 @@ module Tool =
...
+    let interpretFlags  (local: string list) (git: string list) (remote: string list) : string * string * string =
+
+        if not local.IsEmpty then
+            Utils.log "***** local project comparison"
+            let oldFolderPath = local.Item(0)
+            let newFolderPath = local.Item(1)
+            let outFolderPath = local.Item(2)
```
Comment (mschlaipfer) [inline TypeInjections/Tool/Tool.fs]: Just inline these variables. We don't need to assign the RHS to a variable when the variable is just used once. Just use the RHS: `local.Item(0), local.Item(1), local.Item(2)`

File: TypeInjections/Tool/Tool.fs [hunk 7]
```diff
@@ -21,31 +21,38 @@ module Tool =
...
+            Utils.log "***** git commit"
             // location of repo
             let location = Directory.GetCurrentDirectory()
 
-            let commit1 = argvList.Item(1)
-            let commit2 = argvList.Item(2)
-
-            let path = argvList.Item(3)
+            let commit1 = git.Item(0)
+            let commit2 = git.Item(1)
```
Comment (mschlaipfer) [inline TypeInjections/Tool/Tool.fs]: inline `commit1` and `commit2`

File: TypeInjections/Tool/Tool.fs [hunk 8]
```diff
@@ -21,31 +21,38 @@ module Tool =
...
+            let path = git.Item(2)
```
Comment (mschlaipfer) [inline TypeInjections/Tool/Tool.fs]: You can write these as `git.[2]`

File: TypeInjections/Tool/Tool.fs [hunk 9]
```diff
@@ -21,31 +21,38 @@ module Tool =
...
+          git: seq<string>
+          [<Option('r', "remote", Required = false, HelpText = "[github repo url] [new commit] [old commit] [OutputPath]")>]
+          remote: seq<string>
+          [<Option(HelpText = "Prints all messages to standard output.")>]
+          verbose: bool }
 
-            if argvList.Length <> 4 then
-                failwith "usage: program -l OLD_COMMIT NEW_COMMIT OUTPUTFOLDER"
+    // based off command line flags, create and provide path for input/output folders
+    let interpretFlags  (local: string list) (git: string list) (remote: string list) : string * string * string =
```
Comment (mschlaipfer) [inline TypeInjections/Tool/Tool.fs]: This might be nitpicky, but a flag is a boolean command line argument, i.e. one that is either there or not, I'd call ones that have arguments either command line arguments, command line options, or command line parameters.

File: TypeInjections/Tool/Tool.fs [hunk 10]
```diff
@@ -57,41 +64,52 @@ module Tool =
...
             checkoutCommit commit1 url pathOld
             checkoutCommit commit2 url pathNew
 
-        else
-            failwith "flag not found please try again"
 
+            pathOld, pathNew, $"{path}/Output"
+
+        else
+            failwith "unable to read flags"
```
Comment (mschlaipfer) [inline TypeInjections/Tool/Tool.fs]: Do I understand correctly that `local`/`git`/`remote` are mutually exclusive, but at least one of them must be set, with the priority `local`>`git`>`remote`?

If so, please reflect that in the error message here. I don't think `failwith` is the right thing here. You want to print a message that the tool was called incorrectly and gracefully exit.

File: TypeInjections/Tool/Tool.fs [hunk 11]
```diff
@@ -21,31 +21,38 @@ module Tool =
...
+
+        if not local.IsEmpty then
+            Utils.log "***** local project comparison"
+            let oldFolderPath = local.Item(0)
+            let newFolderPath = local.Item(1)
+            let outFolderPath = local.Item(2)
+            oldFolderPath, newFolderPath, outFolderPath
+        else
 
+        if not git.IsEmpty then
```
Comment (mschlaipfer) [inline TypeInjections/Tool/Tool.fs]: why not `elif`

File: TypeInjections/Tool/Tool.fs [hunk 12]
```diff
@@ -57,41 +64,52 @@ module Tool =
...
-        // -r [github repo url] [github commit hash] [github commit hash] [OutputPath]
-        elif flag.Contains("-r") then
-
-            Console.WriteLine("***** remote github")
-
-            if argvList.Length <> 5 then
-                failwith "usage: program -r GITHUB_URL OLD_COMMIT NEW_COMMIT OUTPUTFOLDER"
-
-            let url = argvList.Item(1)
+        else if not remote.IsEmpty then
```
Comment (mschlaipfer) [inline TypeInjections/Tool/Tool.fs]: why not `elif`

File: TypeInjections/Tool/Tool.fs [hunk 13]
```diff
@@ -57,41 +64,52 @@ module Tool =
...
-        let argvList = argv |> Array.toList
-        interpretFlag argvList
+
+        let result =
+            Parser.Default.ParseArguments<options>(argv)
+
+        match result with
+        | :? (Parsed<options>) as parsed -> run parsed.Value
+        | :? (NotParsed<'a>) as notParsed -> ()
+
```
Comment (mschlaipfer) [inline TypeInjections/Tool/Tool.fs:167]: These warnings are worthy of investigation. What's the type of `result`? You might want a `| _ -> ... ` catch all case, but not sure what's the right thing to do here.

File: TypeInjections/Tool/Tool.fsproj [hunk 14]
```diff
@@ -29,6 +29,7 @@
         <PackageReference Include="CommandLineParser.FSharp" Version="2.9.2-ci-210"/>
         <PackageReference Include="DafnyPipeline" Version="3.6.0.40511"/>
         <PackageReference Include="LibGit2Sharp" Version="0.27.0-preview-0182"/>
+        <PackageReference Include="CommandLineParser.FSharp" Version="2.9.2-ci-210" />
```
Comment (mschlaipfer) [inline TypeInjections/Tool/Tool.fsproj]: This dependency is already included on line 29

File: TypeInjections/Tool/Tool.fs (hunk 1 above)
Comment (tschuerl) [reply TypeInjections/Tool/Tool.fs]: Fixed capitalization
`Remote` takes in a github repository from the internet (the repo is not cloned onto the local computer). `git` is when you are in a local repo and want to compare commits of that repository

File: TypeInjections/Tool/Tool.fs (hunk 2 above)
Comment (tschuerl) [reply TypeInjections/Tool/Tool.fs]: Yeah sorry for the confusion, `-d` is when the user wants to use typeCart on two git versions. When comparing git versions, the tool clones two repositories: one at `old commit` and another at `new commit` and then runs typeCart.

 I couldn't figure out whether a user would want to keep these checked out new and old repositories, so I gave them an option to keep these typeCart generated cloned repositories or delete them.

so here is the workflow of the tool when comparing git versions:
clone old project -> clone new project -> run typeCart -> (optional) delete cloned projects

If this is useless, let me know. I can remove this feature!

File: TypeInjections/Tool/Tool.fs (hunk 3 above)
Comment (tschuerl) [reply TypeInjections/Tool/Tool.fs]: I changed help text to `[old project path]` and `[new project path]`

File: TypeInjections/Tool/Tool.fs (hunk 4 above)
Comment (tschuerl) [reply TypeInjections/Tool/Tool.fs]: `local`, `git` and `remote` command need an `outputPath` to write out the typeCart files to. What can I change the name so it's understandable?

File: TypeInjections/Tool/Tool.fsproj (hunk 14 above)
Comment (tschuerl) [reply TypeInjections/Tool/Tool.fsproj]: Fixed

File: TypeInjections/Tool/Tool.fs (hunk 6 above)
Comment (tschuerl) [reply TypeInjections/Tool/Tool.fs]: fixed

File: TypeInjections/Tool/Tool.fs (hunk 7 above)
Comment (tschuerl) [reply TypeInjections/Tool/Tool.fs]: changed lines 53:
`let commit1, commit2 = git.Item(0), git.Item(1)`

File: TypeInjections/Tool/Tool.fs (hunk 8 above)
Comment (tschuerl) [reply TypeInjections/Tool/Tool.fs]: updated all list getters to proper notation

File: TypeInjections/Tool/Tool.fs (hunk 5 above)
Comment (tschuerl) [reply TypeInjections/Tool/Tool.fs]: I made the comment a lot longer but I am not sure if I am getting the message across.
`    (* command line parser has group arguments into what option it is associated with
     (local option has a list of arguments that follow it, as does git and remote).
     If command line option has arguments, we need to organize arguments and run typeCart on them.
     Run typeCart only once per command line call (we can't run a remote and local project on same command line call)
     *)`

File: TypeInjections/Tool/Tool.fs (hunk 9 above)
Comment (tschuerl) [reply TypeInjections/Tool/Tool.fs]: This is helpful, being more precise in the wording will help with comments and variable names

File: TypeInjections/Tool/Tool.fs (hunk 11 above)
Comment (tschuerl) [reply TypeInjections/Tool/Tool.fs]: fixed

File: TypeInjections/Tool/Tool.fs (hunk 12 above)
Comment (tschuerl) [reply TypeInjections/Tool/Tool.fs]: fixed

File: TypeInjections/Tool/Tool.fs (hunk 10 above)
Comment (tschuerl) [reply TypeInjections/Tool/Tool.fs]: Yes, you're correct

I thought this would be a failure, because none of the command line options have any arguments associated with them. We would only reach this `else` statement if none of the flags were correctly called. 

Should I put a help message help the user next time? It would be hard to catch the user's mistake at the point because this `else` statement means the user didn't enter any arguments for any option

File: TypeInjections/Tool/Tool.fs (hunk 13 above)
Comment (tschuerl) [reply TypeInjections/Tool/Tool.fs:167]: I agree, for now I will add `| _ -> ...` and next PR I will handle them with more logic

File: TypeInjections/Tool/Tool.fs (hunk 10 above)
Comment (tschuerl) [reply TypeInjections/Tool/Tool.fs]: @mschlaipfer 
A way to avoid this problem is to use `command line verbs` instead of `command line options`. From the documentation, I think only one verb executes per run. 

Ex of verb:
`typecart local /Users/tschuerl/OldFolder /Users/tschuerl/NewFolder /Users/tschuerl/Output`
Ex of option:
`typecart --local /Users/tschuerl/OldFolder /Users/tschuerl/NewFolder /Users/tschuerl/Output`

File: TypeInjections/Tool/Tool.fs (hunk 1 above)
Comment (mschlaipfer) [reply TypeInjections/Tool/Tool.fs]: Why do we need to support `remote`? Isn't `git` enough?

File: TypeInjections/Tool/Tool.fs (hunk 4 above)
Comment (mschlaipfer) [reply TypeInjections/Tool/Tool.fs]: Why is the output path an argument of these options and not an option by itself?

File: TypeInjections/Tool/Tool.fs (hunk 10 above)
Comment (mschlaipfer) [reply TypeInjections/Tool/Tool.fs]: Yes, that would be nicer.

Comment (mschlaipfer) [reply TypeInjections/Tool/Tool.fs]: And indeed, I think what we want is: 
`typecart local --old /Users/tschuerl/OldFolder --new /Users/tschuerl/NewFolder --out /Users/tschuerl/Output`

Then in code we should be able to refer to the arguments via these names, i.e. you have `local.[0]` or so now, but that should change to something like `local.old` maybe?

File: TypeInjections/Tool/Tool.fs (hunk 1 above)
Comment (tschuerl) [reply TypeInjections/Tool/Tool.fs]: Hira and I thought it would be a convenient feature to have

File: TypeInjections/Tool/Tool.fs (hunk 10 above)
Comment (tschuerl) [reply TypeInjections/Tool/Tool.fs]: Cool, I added a commit where we use CLI verbs instead of options

File: TypeInjections/Tool/Tool.fs [hunk 15]
```diff
@@ -21,77 +57,69 @@ module Tool =
...
 
-        else
-            failwith "flag not found please try again"
+        match result with
 
+        | :? (Parsed<obj>) as command ->
+            match command.Value with
+            | :? LocalOptions as opts -> runLocal opts
+            | :? RemoteOptions as opts -> runRemote opts
+            | :? GitOptions as opts -> runGit opts
```
Comment (mschlaipfer) [inline TypeInjections/Tool/Tool.fs:156]: much better

File: TypeInjections/Tool/Tool.fs [hunk 16]
```diff
@@ -21,77 +57,69 @@ module Tool =
...
 
-    [<EntryPoint>]
-    let main (argv: string array) =
+        | :? (NotParsed<obj>) -> 1 |> ignore
```
Comment (mschlaipfer) [inline TypeInjections/Tool/Tool.fs]: When does this case happen? When we get a verb we don't support? If so we should print the usage help message

File: TypeInjections/Tool/Tool.fs [hunk 17]
```diff
@@ -21,77 +57,69 @@ module Tool =
...
+        let path = remote.outputFolder
+        let pathOld = $"{path}/Old"
+        let pathNew = $"{path}/New"
+        let url = remote.url
+        let commit1 = remote.oldHash
+        let commit2 = remote.newHash
 
+        for a in [ pathOld; pathNew ] do
+            if Directory.Exists(a) then
+                Directory.Delete(a, true)
```
Comment (mschlaipfer) [inline TypeInjections/Tool/Tool.fs]: I think we should have a `--force` option, and if the directory exists and `--force` is not provided we should exit with an error message that the directory exists

File: TypeInjections/Tool/Tool.fs [hunk 18]
```diff
@@ -7,12 +7,48 @@ open LibGit2Sharp
...
+          oldFolder: string
+          [<Option('n', "new", Required = false, HelpText = "folder path of new project")>]
+          newFolder: string
+          [<Option('p', "output", Required = false, HelpText = "folder path to place typeCart files")>]
+          outputFolder: string
+          [<Option('f', "file", Required = false, HelpText = "specify filename you want to run typecart on")>]
+          file: seq<string> }
+
+    [<Verb("remote", HelpText = "[github repo url] [new commit] [old commit] [outputPath]")>]
+    type RemoteOptions =
```
Comment (mschlaipfer) [inline TypeInjections/Tool/Tool.fs]: I think remote should just not exist, at least not as a verb. Let's add an optional `--clone <repo url>` argument to the `git` verb and if it's provided we clone the repository first.

File: TypeInjections/Tool/Tool.fs [hunk 19]
```diff
@@ -7,12 +7,48 @@ open LibGit2Sharp
...
+        { [<Option('o', "old", Required = false, HelpText = "folder path of old project")>]
+          oldFolder: string
+          [<Option('n', "new", Required = false, HelpText = "folder path of new project")>]
+          newFolder: string
+          [<Option('p', "output", Required = false, HelpText = "folder path to place typeCart files")>]
+          outputFolder: string
+          [<Option('f', "file", Required = false, HelpText = "specify filename you want to run typecart on")>]
+          file: seq<string> }
+
+    [<Verb("remote", HelpText = "[github repo url] [new commit] [old commit] [outputPath]")>]
```
Comment (mschlaipfer) [inline TypeInjections/Tool/Tool.fs]: why does this say `github`, can't it be an arbitrary git server?

**Comment by mschlaipfer** [comment]
Please click on "Resolve conversation" when you have addressed the comments

File: TypeInjections/Tool/Tool.fs (hunk 19 above)
Comment (tschuerl) [reply TypeInjections/Tool/Tool.fs]: I know it can clone public repositories hosted on `github` `bitbucket` and `gitlab`. I didn't want people trying to use commands like`ssh://git.amazon.com/pkg/Yucca`

Would a better message be `[https repo url]`?

Comment (mschlaipfer) [reply TypeInjections/Tool/Tool.fs]: just `git URL`

File: TypeInjections/Tool/Tool.fs (hunk 4 above)
Comment (tschuerl) [reply TypeInjections/Tool/Tool.fs]: Now that we use verbs, no longer issue

File: TypeInjections/Tool/Tool.fs (hunk 18 above)
Comment (tschuerl) [reply TypeInjections/Tool/Tool.fs]: I really like this idea, included!

File: TypeInjections/Tool/Tool.fs (hunk 16 above)
Comment (tschuerl) [reply TypeInjections/Tool/Tool.fs]: Yes,
When user has a bad command, it makes a `CommandLine.Error[]` to pinpoint what went wrong

Ex. `typecart dogma git ...`
The not parsed will create `BadVerbError` Object that says `dogma` is not a verb

File: TypeInjections/Tool/Tool.fs (hunk 17 above)
Comment (tschuerl) [reply TypeInjections/Tool/Tool.fs]: I understand now, fixed

File: TypeInjections/Tool/Tool.fs [hunk 20]
```diff
@@ -11,6 +11,33 @@ open CommandLine
...
+
+    [<Verb("git", HelpText = "[old commit] [new commit] [outputPath]")>]
+    type GitOptions =
+        { [<Option('o', "old", Required = true, HelpText = "commit hash of old version")>]
+          oldHash: string
+          [<Option('n', "new", Required = true, HelpText = "commit hash of new version")>]
+          newHash: string
+          [<Option('p', "output", Required = true, HelpText = "folder path to place typeCart files")>]
+          outputFolder: string
+          [<Option('f', "file", Required = false, HelpText = "specify filename you want to run typecart on")>]
```
Comment (mschlaipfer) [inline TypeInjections/Tool/Tool.fs]: the shorthand `f` is duplicated, can we rename this to `entrypoint`?

File: TypeInjections/Tool/Tool.fs [hunk 21]
```diff
@@ -21,91 +48,61 @@ module Tool =
...
-        | :? (Parsed<options>) as parsed -> run parsed.Value
-        | :? (NotParsed<'a>) as notParsed -> Utils.log $"Here is non parsed arguments {notParsed}"
-        | _ as unknown -> Utils.log $"don't understand this argument {unknown}"  
+
+        | :? (Parsed<obj>) as command ->
+            match command.Value with
+            | :? LocalOptions as opts -> runLocal opts
+            | :? GitOptions as opts -> runGit opts
+
+        | :? (NotParsed<obj>) as value-> Utils.log $"finally got to not parsed {value}"
```
Comment (mschlaipfer) [inline TypeInjections/Tool/Tool.fs]: We should print out a usage page you would get from a CLI tool when you run it with `--help` or so. We should also provide such an option, i.e., `typecart --help` that prints the same page

File: TypeInjections/Tool/Tool.fs [hunk 22]
```diff
@@ -21,91 +48,61 @@ module Tool =
...
-                if Directory.Exists(a) then
+        // lib2git needs an empty folder to store the repo clones
+        for a in [ pathOld; pathNew ] do
+            if Directory.Exists(a) then
+                // if user lets us delete existing folder
+                if git.force then
                     Directory.Delete(a, true)
+                // else break program
+                else
+                    failwith $"There is already an {a}, we need empty folder to place git repo. Chose different output folder or chose --force to automatically overwrite data"
```
Comment (mschlaipfer) [inline TypeInjections/Tool/Tool.fs]: typo: choose

in general: please write correct English sentences, don't leave out articles for example

File: TypeInjections/Tool/Tool.fs (hunk 20 above)
Comment (tschuerl) [reply TypeInjections/Tool/Tool.fs]: fixed

File: TypeInjections/Tool/Tool.fs (hunk 21 above)
Comment (tschuerl) [reply TypeInjections/Tool/Tool.fs]: This is built in feature of `CommandLine.Parser` All the `HelpText` is what gets written out when user asks for help
Ex. `typecart git --help`
Ouptut:

```
  -o, --old           Required. commit hash of old version

  -n, --new           Required. commit hash of new version

  -p, --output        Required. folder path to place typeCart files

  -e, --entrypoint    specify filename you want to run typecart on

  -c, --clone         Give a git URL

  -f, --force         Overwrite the New and Old folders in Output folder

  --help              Display this help screen.

  --version           Display version information.
```

File: TypeInjections/Tool/Tool.fs (hunk 22 above)
Comment (tschuerl) [reply TypeInjections/Tool/Tool.fs]: Ugh that was lazy, sorry

File: TypeInjections/Tool/Tool.fs (hunk 21 above)
Comment (tschuerl) [reply TypeInjections/Tool/Tool.fs]: I noticed the capitalization is off, will fix shortly

Comment (tschuerl) [reply TypeInjections/Tool/Tool.fs]: Fixed

File: TypeInjections/Tool/Tool.fs [hunk 23]
```diff
@@ -21,77 +48,64 @@ module Tool =
...
-            failwith "flag not found please try again"
+        match result with
+        | :? (Parsed<obj>) as command ->
+            match command.Value with
+            | :? LocalOptions as opts -> runLocal opts
+            | :? GitOptions as opts -> runGit opts
+            | _ -> 0 |> ignore
 
+        | :? (NotParsed<obj>) as value -> 0 |> ignore
+        | _ -> 0 |> ignore
```
Comment (mschlaipfer) [inline TypeInjections/Tool/Tool.fs]: shouldn't all of these `0 |> ignore` print the help message and exit the program? Minor note, `0 |> ignore` could probably be simplified to `()`

File: TypeInjections/Tool/Tool.fs [hunk 24]
```diff
@@ -7,12 +7,39 @@ open LibGit2Sharp
...
+          [<Option('c', "clone", Required = false, HelpText = "input a git URL")>]
+          url: string
+          [<Option('f', "force", Required = false, HelpText = "overwrite the Old and New folders in the Output folder")>]
+          force: bool }
+
     //NOTE 'location' refers to either the github url OR path to repo on local computer
     let checkoutCommit (hash: string) (location: string) (outputPath: string) : unit =
-        Utils.log $"path to repo is {outputPath}"
 
         Repository.Clone(location, outputPath) |> ignore
```
Comment (mschlaipfer) [inline TypeInjections/Tool/Tool.fs]: this should only be done when `--clone <url>` is set, is that true, it's hard to read the diff

File: TypeInjections/Tool/Tool.fs (hunk 23 above)
Comment (tschuerl) [reply TypeInjections/Tool/Tool.fs]: The `CommandLine.Parser` will automatically throw an error message, that is why I ignored the `NotParsed` sequence. I will simplify to `()`

File: TypeInjections/Tool/Tool.fs (hunk 24 above)
Comment (tschuerl) [reply TypeInjections/Tool/Tool.fs]: Ok, I will print out the output path for `--clone <url>`

File: TypeInjections/Tool/Tool.fs [hunk 25]
```diff
@@ -7,91 +7,102 @@ open LibGit2Sharp
...
+
     //NOTE 'location' refers to either the github url OR path to repo on local computer
     let checkoutCommit (hash: string) (location: string) (outputPath: string) : unit =
-        Utils.log $"path to repo is {outputPath}"
-
         Repository.Clone(location, outputPath) |> ignore
         // get Repository object of whatever was cloned above
         let newRepo = new Repository(outputPath)
         // checkout repo at given commit
         let commitRepo = newRepo.Lookup<Commit>(hash)
```
Comment (mschlaipfer) [inline TypeInjections/Tool/Tool.fs:66]: This would fail if the current directory isn't from git, we should handle that failure

File: TypeInjections/Tool/Tool.fs [hunk 26]
```diff
@@ -7,91 +7,102 @@ open LibGit2Sharp
...
+          [<Option('c', "clone", Required = false, HelpText = "input a git URL")>]
+          url: string
+          [<Option('f', "force", Required = false, HelpText = "overwrite the Old and New folders in the Output folder")>]
+          force: bool }
+
     //NOTE 'location' refers to either the github url OR path to repo on local computer
     let checkoutCommit (hash: string) (location: string) (outputPath: string) : unit =
-        Utils.log $"path to repo is {outputPath}"
-
         Repository.Clone(location, outputPath) |> ignore
```
Comment (mschlaipfer) [inline TypeInjections/Tool/Tool.fs]: Please add a comment here, also add a link to the library's documentation, if possible.

File: TypeInjections/Tool.Test/ToolTests.fs [hunk 27]
```diff
@@ -0,0 +1,148 @@
...
+        (* TODO ask Matthias how to replicate local git repo in unit testing (git wanted me to make a submodule)
+        let outFolder = $"{resourcePath}/Git/BadLocation/Generated/Output"
+        // ensure output directory is empty (in Tool.fs, other method handles this, checkout needs this to be done manually)
+        if Directory.Exists(outFolder) then
+            Directory.Delete(outFolder, true)
+
+        let exnNoDafny =
+            Assert.Catch(fun () -> Tool.checkoutCommit "20e447" $"{resourcePath}/Git/BadLocation/RepoNoDafny" outFolder)
+
+        Console.WriteLine $"Exn was: {exnNoDafny.Message}" *)
```
Comment (tschuerl) [inline TypeInjections/Tool.Test/ToolTests.fs]: I know we don't comment code, but I wanted to ask you about testing `typecart git` with a local repository. I made a repository inside `Tool/Resources` but when I tried to commit a change, I got an error. The error message wanted me to add a `git submodule`

File: TypeInjections/Tool/Tool.fs (hunk 25 above)
Comment (tschuerl) [reply TypeInjections/Tool/Tool.fs:66]: Handled, and useful error message is given

File: TypeInjections/Tool/Tool.fs (hunk 26 above)
Comment (tschuerl) [reply TypeInjections/Tool/Tool.fs]: I added link to libraries in the README, will also add here:
[CommandLineParser.fsharp](https://github.com/commandlineparser/commandline) Popular .NET command line parsing library
[Lib2GitSharp](https://github.com/libgit2/libgit2sharp/wiki) and [Lib2Git Manual](https://libgit2.org/docs/guides/101-samples/) Easy to use .NET wrapper for git commands

File: TypeInjections/Tool.Test/Resources/Git/BadLocation/file2.dfy [hunk 28]
```diff
@@ -0,0 +1,13 @@
+
+module S {
+
+
+
+// BOOM NOW YOU SEE ME
```
Comment (mschlaipfer) [inline TypeInjections/Tool.Test/Resources/Git/BadLocation/file2.dfy:6]: unnecessary whitespace above and the comment doesn't help

File: TypeInjections/Tool.Test/ToolTests.fs [hunk 29]
```diff
@@ -0,0 +1,148 @@
...
+open TestUtils
+open Tool
+
+
+module IOTests =
+    open NUnit.Framework
+
+
+
+    [<Test>]
```
Comment (mschlaipfer) [inline TypeInjections/Tool.Test/ToolTests.fs:11]: Please keep the amount of whitespace to a single empty line. Did you run the formatter on this?

File: TypeInjections/Tool.Test/ToolTests.fs (hunk 27 above)
Comment (mschlaipfer) [reply TypeInjections/Tool.Test/ToolTests.fs]: I think we'd need to mock the git repository, so the library thinks there is one, but in reality it's just a mock. However, I think we can defer those tests for now. Just add a comment `// TODO: test typecart git --clone <URL>`

Comment (mschlaipfer) [reply TypeInjections/Tool.Test/ToolTests.fs]: BTW: the same is true for all the git commands above, we should not test these with real git repositories. Please add a TODO instead of using your hard-coded personal reposiotry. We can look into it later.

This might be relevant, if we have time to get back to it:
https://github.com/libgit2/libgit2sharp/blob/master/LibGit2Sharp.Tests/MockingFixture.cs

File: TypeInjections/Tool.Test/Resources/Git/BadLocation/file2.dfy (hunk 28 above)
Comment (tschuerl) [reply TypeInjections/Tool.Test/Resources/Git/BadLocation/file2.dfy:6]: This is apart of the remote repository test that we want to remove in favor of mockup. I will delete all these files

File: TypeInjections/Tool.Test/ToolTests.fs (hunk 29 above)
Comment (tschuerl) [reply TypeInjections/Tool.Test/ToolTests.fs:11]: Yes I ran the formatter, apparently doesn't help with this. Will keep white space to single line

File: TypeInjections/Tool.Test/ToolTests.fs [hunk 30]
```diff
@@ -0,0 +1,117 @@
...
+    // raise exception for bad git URL
+    let GitBadURLTest () =
+        Assert.Catch
+            (fun () ->
+                Tool.checkoutCommit
+                    "0f836ae"
+                    "https://github.com/awslabs/IMAGINARY_REPO.git"
+                    $"{resourcePath}/Git/BadURLTest/Output")
+        |> ignore
```
Comment (tschuerl) [inline TypeInjections/Tool.Test/ToolTests.fs:82]: I saved this git test because the url doesn't actually exist. However, I can still delete it

File: TypeInjections/Tool.Test/ToolTests.fs [hunk 31]
```diff
@@ -0,0 +1,117 @@
...
+        // no git repo
+        let exn =
+            Assert.Catch
+                (fun () ->
+                    Tool.checkoutCommit
+                        "0f836ae"
+                        $"{resourcePath}/Git/BadLocation"
+                        $"{resourcePath}/Git/BadLocation/Generated")
+
+        Console.WriteLine $"Exn was: {exn.Message}"
```
Comment (tschuerl) [inline TypeInjections/Tool.Test/ToolTests.fs:96]: I saved git test because this isn't a real repo, just making sure CLI handles error properly

**Comment by ruijiefang** [comment]
adding myself as a reviewer as I requested a change

File: TypeInjections/Tool.Test/ToolTests.fs [hunk 32]
```diff
@@ -0,0 +1,117 @@
...
+        Assert.AreEqual(outputFilenames, expectedFilenames)
+
+    //TODO test long commit hashes
+
+    //TODO test short commit hashes
+    
+    //TODO test commits not from "main"
+    
+    //TODO test typecart git --clone <URL>
+
```
Comment (tschuerl) [inline TypeInjections/Tool.Test/ToolTests.fs:72]: Removed the git repo tests

**Review by mschlaipfer — APPROVED** [review]

**Comment by mschlaipfer** [comment]
> adding myself as a reviewer as I requested a change

@ruijiefang I missed your comment, and I wasn't blocked for merging this. Feel free to add comments and we'll address them in another PR.
